compiler/rustc_middle/src/query/mod.rs

```diff
@@ -2109,12 +2109,12 @@ rustc_queries! {
         separate_provide_extern
     }
 
-    query permits_uninit_init(key: TyAndLayout<'tcx>) -> bool {
-        desc { "checking to see if `{}` permits being left uninit", key.ty }
+    query permits_uninit_init(key: (ty::ParamEnv<'tcx>, TyAndLayout<'tcx>)) -> bool {
```

Comment:
This should use `ParamEnvAnd` if possible. It has an optimization to avoid some query invalidation.

Reply:
`ParamEnvAnd` requires `T: TypeVisitable`, which is not implemented by `TyAndLayout`. I think the implementation of that trait would be straightforward, but I'm not sure where to put it.

Reply:
There is a ty::structural_impls file for impls we do not know where to put.